### PR TITLE
Added changes for observer and singleton design patterns

### PR DESCRIPTION
**Pull request by karenathaide**

### CODE REVIEW

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java [hunk 1]
```diff
@@ -44,17 +49,39 @@
...
+	public MainController getMainController() {
+		return mainController;
+	}
+	
+	/**
+	 * This method sets the MainController.
+	 */
+	public void setMainController(MainController mainCtrl) {
+		this.mainController = mainCtrl;
+	}
```
Comment (NeelimaNair) [inline src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: The getter and setter for main controller is there in the BaseController. We don't need to put them here again.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java [hunk 2]
```diff
@@ -44,17 +49,39 @@
 	private int price=0;
 	/**Identifier of the selected drink.*/
 	private int selection=-1;
+	private MainController mainController;
```
Comment (NeelimaNair) [inline src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: The mainController instance is already there in the BaseController. Not required to put here again.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java [hunk 3]
```diff
@@ -44,17 +49,39 @@
...
 		coinReceiver=new CoinReceiver(this);
 		changeGiver=new ChangeGiver(this);
 	}
+	
+	public static synchronized TransactionController getInstance(){
+        if (uniqueInstance == null) {
+        	uniqueInstance = new TransactionController();         
+        } 
+        return uniqueInstance;
+    }
```
Comment (NeelimaNair) [inline src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: Both the constructor and getInstance methods might be needing the MainController instance to be sent in as input parameter and set with the transactionController when it is being instantiated.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java [hunk 4]
```diff
@@ -147,8 +178,6 @@ public void completeTransaction(){
 		}
 		coinReceiver.storeCash();
 		dispenseCtrl.allowSelection(true);
-		
-		refreshMachineryDisplay();
```
Comment (NeelimaNair) [inline src/sg/edu/nus/iss/vmcs/customer/TransactionController.java:151]: Why is this removed? Where will it be invoked now within the TransactionController?

**Review by NeelimaNair — CHANGES_REQUESTED** [review]
Overall looks good. Only minor comments and changes required.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java (hunk 3 above)
Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: getInstance method cannot take a parameter.  This was the problem I faced.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java (hunk 4 above)
Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java:151]: It is invoked in the observer update method in MachineryController.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java (hunk 1 above)
Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: The problem seems to be the project not seeing the BaseController file.

Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: Fixed.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java (hunk 2 above)
Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: Fixed.

File: src/sg/edu/nus/iss/vmcs/customer/TransactionController.java (hunk 3 above)
Comment (karenathaide) [reply src/sg/edu/nus/iss/vmcs/customer/TransactionController.java]: Fixed inherently by the previous comment fixes.

**Review by NeelimaNair — APPROVED** [review]
Looks good. Approving the merge.
